In [1]:
import pandas as pd
import matplotlib.pyplot as plt

# Read the CSV and Perform Basic Data Cleaning

In [2]:
df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')
# Drop the null rows
df = df.dropna()
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

# Select your features (columns)

In [3]:
# Set features. This will also be used as your x values.
# selected_features = df[['names', 'of', 'selected', 'features', 'here']]
selected_features = df

# Create a Train Test Split

Use `koi_disposition` for the y values

In [4]:
# Assign x and y values
X = df.drop("koi_disposition", axis=1)
y = df["koi_disposition"]

print(X.shape, y.shape)

(6991, 40) (6991,)


In [5]:
# Import Dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler


# Split data into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=115, stratify=y)

In [6]:
X_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
2700              0              0              0              0   31.804760   
21                0              0              0              0    4.280964   
5040              0              1              1              0    0.829019   
2795              0              0              1              0    1.387838   
1765              0              0              0              0   14.383227   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
2700         0.000091        -0.000091    190.25413           0.00224   
21           0.000006        -0.000006    171.89659           0.00115   
5040         0.000002        -0.000002    131.46667           0.00401   
2795         0.000006        -0.000006    132.11351           0.00318   
1765         0.000086        -0.000086    140.57648           0.00530   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
2700          -0.00224  ...            -128      4.235           0.143   
21            -0.00115  ...            -169      4.564           0.032   
5040          -0.00401  ...            -255      4.271           0.124   
2795          -0.00318  ...            -219      4.442           0.054   
1765          -0.00530  ...            -135      4.249           0.137   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
2700          -0.117     1.356          0.212         -0.259  287.62082   
21            -0.168     0.831          0.207         -0.069  294.26581   
5040          -0.186     1.328          0.408         -0.220  296.03485   
2795          -0.216     1.044          0.335         -0.112  295.64212   
1765          -0.112     1.291          0.216         -0.216  290.08276   

            dec  koi_kepmag  
2700  43.141647      13.178  
21    49.314091      15.356  
5040  45.490822      14.661  
2795  48.773071      15.167  
1765  50.863480      13.312  

[5 rows x 40 columns]

# Pre-processing

Scale the data using the MinMaxScaler and perform some feature selection

In [7]:
# Scale the data
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Train the Model



In [8]:
# Create the Random Forest Model
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf = rf.fit(X_train_scaled, y_train)

In [9]:
print(f"Training Data Score: {rf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {rf.score(X_test_scaled, y_test)}")

Training Data Score: 1.0
Testing Data Score: 0.8907322654462243


In [10]:
# Random Forests in sklearn will automatically calculate feature importance
importances = rf.feature_importances_
importances

array([0.10493894, 0.07816168, 0.11137466, 0.03287196, 0.02022085,
       0.02075999, 0.01640427, 0.01426515, 0.02425956, 0.01860485,
       0.01934175, 0.01014886, 0.01000429, 0.02129515, 0.03306089,
       0.03295789, 0.02057538, 0.01335091, 0.01436844, 0.04650248,
       0.03662917, 0.02147508, 0.01658801, 0.01177511, 0.02076884,
       0.01018816, 0.06465583, 0.0026524 , 0.00942565, 0.0287305 ,
       0.02203845, 0.00894239, 0.00850378, 0.01036633, 0.00877143,
       0.0110187 , 0.00852364, 0.01338762, 0.0111281 , 0.01096287])

In [11]:
# Sort the features by their importance
sorted(zip(rf.feature_importances_, selected_features), reverse=True)

[(0.11137465906368486, 'koi_fpflag_ss'),
 (0.1049389431047826, 'koi_disposition'),
 (0.07816168094655228, 'koi_fpflag_nt'),
 (0.06465583388418827, 'koi_insol_err2'),
 (0.04650247524789392, 'koi_depth_err2'),
 (0.036629167282281404, 'koi_prad'),
 (0.03306088654517188, 'koi_duration'),
 (0.0329578881829076, 'koi_duration_err1'),
 (0.03287196335103194, 'koi_fpflag_co'),
 (0.02873049677158351, 'koi_steff'),
 (0.02425955873918812, 'koi_time0bk'),
 (0.022038451661523756, 'koi_steff_err1'),
 (0.021475081108284284, 'koi_prad_err1'),
 (0.021295148010125064, 'koi_impact_err2'),
 (0.020768843165358278, 'koi_insol'),
 (0.020759989424587684, 'koi_period'),
 (0.020575381296357262, 'koi_duration_err2'),
 (0.020220845364379392, 'koi_fpflag_ec'),
 (0.01934175236942166, 'koi_time0bk_err2'),
 (0.01860484857164134, 'koi_time0bk_err1'),
 (0.016588005343421575, 'koi_prad_err2'),
 (0.01640426843507561, 'koi_period_err1'),
 (0.014368442096669704, 'koi_depth_err1'),
 (0.01426515358986645, 'koi_period_err2'),
 

# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [12]:
# Create the GridSearchCV model
from sklearn.model_selection import GridSearchCV
param_grid = {'n_estimators': [250, 300, 350],
              'max_depth': [125, 150, 175]}
grid = GridSearchCV(rf, param_grid, verbose=3)

In [13]:
# Fit the Random Forest model using the grid search estimator. 
# This will take the Random Forest model and try each combination of parameters
grid.fit(X_train_scaled, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ..... max_depth=125, n_estimators=250, score=0.901, total=   2.7s
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    2.6s remaining:    0.0s


[CV] ..... max_depth=125, n_estimators=250, score=0.891, total=   2.7s
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    5.3s remaining:    0.0s


[CV] ..... max_depth=125, n_estimators=250, score=0.895, total=   2.7s
[CV] max_depth=125, n_estimators=250 .................................
[CV] ..... max_depth=125, n_estimators=250, score=0.895, total=   2.7s
[CV] max_depth=125, n_estimators=250 .................................
[CV] ..... max_depth=125, n_estimators=250, score=0.892, total=   2.7s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.898, total=   3.2s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.894, total=   3.2s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.903, total=   3.2s
[CV] max_depth=125, n_estimators=300 .................................
[CV] ..... max_depth=125, n_estimators=300, score=0.893, total=   3.2s
[CV] max_depth=125, n_estimators=300 .................................
[CV] .

[Parallel(n_jobs=1)]: Done  45 out of  45 | elapsed:  2.4min finished


GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'max_depth': [125, 150, 175],
                         'n_estimators': [250, 300, 350]},
             verbose=3)

In [14]:
# List the best parameters for this dataset
print(f"Best Parameters: {grid.best_params_}")

# List the best score
print(f"Best Score: {grid.best_score_}")

Best Parameters: {'max_depth': 150, 'n_estimators': 350}
Best Score: 0.8968144870796614


In [15]:
print(f"Training Grid Score: {grid.score(X_train_scaled, y_train)}")
print(f"Testing Grid Score: {grid.score(X_test_scaled, y_test)}")

Training Grid Score: 1.0
Testing Grid Score: 0.8913043478260869


In [16]:
# Make predictions with the hypertuned model
predictions = grid.predict(X_test_scaled)

In [17]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

                precision    recall  f1-score   support

     CANDIDATE       0.82      0.73      0.77       422
     CONFIRMED       0.78      0.84      0.81       450
FALSE POSITIVE       0.98      1.00      0.99       876

      accuracy                           0.89      1748
     macro avg       0.86      0.85      0.86      1748
  weighted avg       0.89      0.89      0.89      1748



# Save the Model

In [18]:
# Save the model 

import joblib
filename = 'LCortez_RF.sav'
joblib.dump(rf, filename)

['LCortez_RF.sav']

# Conclusion

The f1-score of FALSE POSITIVE for Random Forest model is closer to 1 (0.99) meaning it can predict FALSE POSITIVE well and it's reliable.

The Grid score from the hyperparameter tuning boost the Random Forest score higher Training Data Score: 1.0 remains 1.0  Testing Data Score: 0.8907322654462243 changes to 0.8913043478260869 Best Score: 0.8968144870796614

However, Random Forest model's best score seems better than SVM model when comparing the scores.